sumdbaudit/witness/cmd/feeder/main.go

```diff
@@ -0,0 +1,295 @@
+	witnessKey   = flag.String("wk", "", "The public key of the witness")
+	logID        = flag.String("lid", "", "The ID of the log within the witness")
+	pollInterval = flag.Duration("poll", 10*time.Second, "How quickly to poll the sumdb to get updates")
+
+	rf = &compact.RangeFactory{
+		Hash: rfc6962.DefaultHasher.HashChildren,
+	}
+)
+
+const (
+	tileHeight    = 8
+	leavesPerTile = 1 << tileHeight
+)
+
+func main() {
+	flag.Parse()
+	ctx := context.Background()
+	sdb := audit.NewSumDB(tileHeight, *vkey)
+	var w http.Witness
+	if wURL, err := url.Parse(*witness); err != nil {
+		glog.Exitf("Failed to parse witness URL: %v", err)
+	} else {
+		w = http.Witness{
+			URL:      wURL,
+			Verifier: mustCreateVerifier(*witnessKey),
+		}
+	}
+
+	wcp := &log.Checkpoint{}
+	if wcpRaw, err := w.GetLatestCheckpoint(ctx, *logID); err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			glog.Exitf("Failed to get witness checkpoint: %v", err)
+		}
+	} else {
+		n, err := note.Open(wcpRaw, note.VerifierList(w.Verifier))
+		if err != nil {
+			glog.Exitf("Failed to open CP: %v", err)
+		}
+
+		_, err = wcp.Unmarshal([]byte(n.Text))
+		if err != nil {
+			glog.Exitf("Failed to unmarshal CP: %v", err)
+		}
+	}
+
+	feeder := feeder{
+		wcp: wcp,
+		sdb: sdb,
+		w:   w,
+	}
+
+	tik := time.NewTicker(*pollInterval)
+	for {
+		glog.V(2).Infof("Tick: start feedOnce (witness size %d)", feeder.wcp.Size)
+		if err := feeder.feedOnce(ctx); err != nil {
+			glog.Exitf("Failed to feed: %v", err)
+		}
+		glog.V(2).Infof("Tick: feedOnce complete (witness size %d)", feeder.wcp.Size)
+
+		select {
+		case <-ctx.Done():
+			return
+		case <-tik.C:
+		}
+	}
+}
+
+// feeder encapsulates the main logic and state of the feeder.
+// The residual logic outside of this should simply be initialization and
+// orchestration with a timer.
+// Note that feeder maintains its own state that represents the witness state.
+// This is optimized for the case where this is the only feeder for this log to
+// the witness. If the witness state is updated outwith this feeder, then the
+// number of successful updates from this feeder will drop. On the other hand,
+// if this is the only feeder then it avoids making requests to get the latest
+// checkpoint from the witness when it isn't changing.
+type feeder struct {
+	wcp *log.Checkpoint
+	sdb *audit.SumDBClient
+	w   http.Witness
+}
+
+// feedOnce gets the latest checkpoint from the SumDB server, and if this is more
+// recent than the witness state then it will construct a compact range proof by
+// requesting the minimal set of tiles, and then update the witness with the new
+// checkpoint. Finally, it will update the feeder's view of the witness state.
+func (f *feeder) feedOnce(ctx context.Context) error {
+	sdbcp, err := f.sdb.LatestCheckpoint()
+	if err != nil {
+		return fmt.Errorf("failed to get latest checkpoint: %v", err)
+	}
+	if int64(f.wcp.Size) >= sdbcp.N {
```

Comment:
If the witness' checkpoint is strictly bigger than the latest sumDB checkpoint then something has really gone wrong, right?  Is it worth treating this case separately and exiting/logging some kind of sadness?

Reply:
In distributed systems it could happen that a "server" (actually a collection of server instances) is not internally consistent and thus making 2 quick requests gets you a fresh answer and a stale answer. Eventually they will become consistent, and nothing nefarious is happening here.

If anything nefarious is happening, then it will eventually get caught out by the consistency check when the tree next grows.

I could separate this out and Warn in the case of a size regression as a curio, but I wouldn't make it bail at this point. If you're still attached to the printing of a stronger message then comment here and I'll make it so.